File: WTRSimlib/src/main/java/simulation/launch/SocketHandler.java
package simulation.launch;

import java.io.IOException;
import java.net.DatagramPacket;
import java.net.DatagramSocket;
import java.net.InetAddress;
import java.net.SocketException;
import simulation.launch.UnityVerifier;

// Credit to the University of Northampton, UK
// http://www.eng.northampton.ac.uk/~espen/CSY2026/JavaServerCSClient.htm
// CSY2026 Modern Networks

/**
 * Class that handles communication with Unity.
 */
class SocketHandler implements Runnable  {

    Runnable unityVerifier;

    public int sendPort = 4513;
    public int receivePort = 4512;

    public DatagramSocket receiveSocket;
    public DatagramSocket sendSocket;
    InetAddress address;

    public String outboundString = "hello Unity";
    public String received;

    volatile boolean kill = false;
    int lastHeartBeat = -1;
    int heartBeat = 0;
    Boolean heartStopped = false;

    SocketHandler() {
        address = InetAddress.getLoopbackAddress();

        try {
            sendSocket = new DatagramSocket();
            receiveSocket = new DatagramSocket(receivePort);
            Runtime.getRuntime().addShutdownHook(new Thread() {
                public void run() {
                    receiveSocket.close();
                    sendSocket.close();
                }
            });
        } catch(SocketException e) {
            e.printStackTrace();
            kill = true;
        }

        unityVerifier = new UnityVerifier(this);
    }

    public void run() {
        unityVerifier.run();
   
        //Outbound
        new Thread(() -> { 
            while (!kill) {
                try {
                    Thread.sleep(10);
                } catch (InterruptedException e) {
                    e.printStackTrace();
                }
                try {
                    byte[] buffer = new byte[256];
                    buffer = outboundString.getBytes();
                    DatagramPacket outboundDp = 
                        new DatagramPacket(
                        buffer, 
                        buffer.length, 
                        address, 
                        sendPort
                    );
    
                    sendSocket.send(outboundDp);

                } catch (IOException e) {
                    e.printStackTrace();
                }
            }
        }).start();

        //Inbound
        new Thread(() -> { 
            while (!kill) { 
                try {
                    Thread.sleep(10);
                } catch(InterruptedException e) {
                    e.printStackTrace();
                }

                try {
                    lastHeartBeat = heartBeat;
                    DatagramPacket inboundDp;
                    byte[] buffer = new byte[256];
                    inboundDp = new DatagramPacket(buffer, buffer.length);
                    receiveSocket.receive(inboundDp);
                    heartBeat++;
                    received = new String(inboundDp.getData());
                    System.out.println("packet received: " + received);
                } catch(IOException e) {
                    e.printStackTrace();
                }

                if (heartStopped) { 
                    System.out.print("\nTimed out. Aborting... \n\n");
                    kill = true;
                }

                heartBeat = (heartBeat >= 1000000) ? 0 : heartBeat;
                heartStopped = (heartBeat == lastHeartBeat) ? true 
                    : heartStopped;
            }
        }).start();         
    }
}

File: WTRSimUnity/Assets/Scripts/SocketHandler.cs
﻿using System;
using System.Net;
using System.Net.Sockets;
using UnityEngine;
using UnityEditor;
using UnityEditor.PackageManager;
using System.Collections;
using UnityEngine.Assertions;
using System.Threading;
using System.Linq;
using System.Text;

//Singleton needed on game object

// Credit to the University of Northampton, UK
// http://www.eng.northampton.ac.uk/~espen/CSY2026/JavaServerCSClient.htm
// CSY2026 Modern Networks

/**
 * Class that handles communication with Java
 */
[InitializeOnLoad]
public class SocketHandler: MonoBehaviour
{
    int clientPort = 4513;
    int serverPort = 4512;

    Socket sendSocket;
    UdpClient receiveSocket;
    IPEndPoint clientEP;
    IPAddress serverIPAddress;
    IPEndPoint serverEP;

    Boolean dependentMode; //If dependentMode is true, Unity is running dependently to WTRSimlib
    volatile Boolean kill = false;
	
	Boolean initialized = false;

    SocketHandler()
    {
       
        sendSocket = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp);
        receiveSocket = new UdpClient(clientPort);
        clientEP = new IPEndPoint(IPAddress.Any, clientPort);
        sendSocket.SendTimeout = 10000;
        receiveSocket.Client.ReceiveTimeout = 10000;

        try
        {
            serverIPAddress = IPAddress.Parse("127.0.0.1");
            serverEP = new IPEndPoint(serverIPAddress, serverPort);

            //Tell java that Unity has launched
            byte[] buffer = new byte[256];
            buffer = Encoding.ASCII.GetBytes("");
            sendSocket.SendTo(buffer, serverEP);

            //Verify that Java knows Unity is alive
            byte[] inboundBytes = receiveSocket.Receive(ref clientEP);

            Debug.Log("Handshake Success!");
            dependentMode = true;
			EditorApplication.playModeStateChanged += playModeChanged;
        }
        catch ( SocketException e )
        {
            Debug.Log("Running Independent to RobotCode");
            dependentMode = false;
        }
		
       EditorApplication.update += Update;
    }

    private void playModeChanged(PlayModeStateChange state)
    {
        //Tell java that playmode has been reached
        byte[] buffer = new byte[ 256 ];
        buffer = Encoding.ASCII.GetBytes("");
        sendSocket.SendTo(buffer, serverEP);

        if(dependentMode)
        {
            //Inbound
            new Thread(() =>
            {
                while ( !kill )
                {
                    Thread.Sleep(5);
                    try
                    {
                        byte[] inboundBytes = receiveSocket.Receive(ref clientEP);
                        Debug.Log("packet received");
                    }
                    catch ( SocketException e )
                    {
                        Debug.Log("Timed Out. Aborting...");
                        kill = true;
                    }
                }
            }).Start();

            //Outbound
            new Thread(() =>
            {
                while ( !kill )
                {
                    Thread.Sleep(30);
                    try
                    {
                        byte[] message = Encoding.ASCII.GetBytes("");
                        sendSocket.SendTo(message, serverEP);
                    }
                    catch ( SocketException e )
                    {
                        Debug.LogError(e.Message + " " + e.StackTrace);
                    }
                }
            }).Start();
        }
    }

    private void Update()
    {
        if ( !initialized && dependentMode )
        {
            EditorApplication.EnterPlaymode();
            initialized = true;
        }

        if ( dependentMode && !EditorApplication.isPlaying )
        {
            receiveSocket.Close();
            sendSocket.Close();
        }

        if ( kill )
        {
            killProcesses();
        }
    }

    void OnApplicationQuit()
    {
        killProcesses();
    }

    void killProcesses()
    {
        AppDomain.CurrentDomain.ProcessExit += (obj, eArg) =>
        {
            kill = true;
        };

        receiveSocket.Close();
        sendSocket.Close();
        EditorApplication.Exit(0);
    }
}


